src/Agent.Worker/TestResults/TestRunPublisher.cs

```diff
@@ -0,0 +1,306 @@
+                completedDate: _testRunData.CompleteDate,
+                state: TestRunState.Completed.ToString()
+                );
+            Task<TestRun> sendAsycn = _httpClient.UpdateTestRunAsync(_projectName, _testRun.Id, updateModel);
+            sendAsycn.Wait();
+
+            // Uploading run level attachments, only after run is marked completed;
+            // so as to make sure that any server jobs that acts on the uploaded data (like CoverAn job does for Coverage files)  
+            // have a fully published test run results, in case it wants to iterate over results 
+            if (publishAttachmentsAsArchive)
+            {
+                UploadTestRunAttachmentsAsArchive();
+            }
+            else
+            {
+                UploadTestRunAttachmentsIndividual();
+            }
+
+            _testRun = sendAsycn.Result;
+            return _testRun.WebAccessUrl;
+        }
+
+        /// <summary>
+        /// Converts the given results file to TestRunData object
+        /// </summary>
+        /// <param name="filePath">File path</param>
+        /// <returns>TestRunData</returns>
+        public TestRunData ReadResultsFromFile(string filePath)
+        {
+            return _resultReader.ReadResults(filePath, _runContext);
+        }
+
+        /// <summary>
+        /// Converts the given results file to TestRunData object
+        /// </summary>
+        /// <param name="filePath">File path</param>
+        /// <param name="runName">Run Name</param>
+        /// <returns>TestRunData</returns>
+        public TestRunData ReadResultsFromFile(string filePath, string runName)
+        {
+            _runContext.RunName = runName;
+            return _resultReader.ReadResults(filePath, _runContext);
+        }
+        #endregion
+
+        private void UploadTestRunAttachmentsAsArchive()
+        {
+            // Do not upload duplicate entries 
+            var attachedFiles = UniqueTestRunFiles;
+            try
+            {
+                var tempDirectory = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+                Directory.CreateDirectory(tempDirectory);
+                var zipFile = Path.Combine(tempDirectory, "TestResults_" + _testRun.Id + ".zip");
+
+                File.Delete(zipFile); //if there's already file. remove silently without exception
+                zipper.CreateZipFile(zipFile, attachedFiles);
+                CreateTestRunAttachment(zipFile);
+            }
+            catch (Exception ex)
+            {
+                _executionContext.Warning(StringUtil.Loc("UnableToArchiveResults", ex));
+                UploadTestRunAttachmentsIndividual();
+            }
+        }
+
+        private void UploadTestRunAttachmentsIndividual()
+        {
+            _executionContext.Debug("Uploading test run attachements individually");
+            // Do not upload duplicate entries 
+            var attachedFiles = UniqueTestRunFiles;
```

Comment:
Another style note: we use var only on the right hand side if the type is clear. If its not we explicitly define the type.